include/tvm/target/se_scope.h

```diff
@@ -0,0 +1,333 @@
+ * simply insert "device_copy" operators wherever \p SEScopes are not exactly pointwise equal.
+ * We may revisit this in the future as the work on memory pools matures.
+ *
+ * Joining and Defaulting
+ * ----------------------
+ * It is possible to 'join' two \p SEScopes to yield the most constrained \p SEScope which agrees
+ * with both join arguments. Eg:
+ * \code
+ * Join((kDLCPU, -1, "llvm", ""), (kInvalidDeviceType, 3, null, "global))
+ *   => (kDLCPU, 3, "llvm", "global")
+ * Join((kDLCPU, -1, "llvm", ""), (kInvalidDeviceType, 3, null, "local))
+ *   => null (no join possible)
+ * \endcode
+ *
+ * Related to 'join' is 'default', which only takes constrained fields from the rhs when the
+ * lhs is unconstrained:
+ * \code
+ * Default(kDLCPU, -1, "llvm", "local"), (kDLCPU, 3, null, "global"))
+ *   => (kDLCPU, 3, "llvm", "local")
+ * \endcode
+ *
+ * These operations are needed during device planning.
+ *
+ */
+class SEScopeNode : public Object {
+ public:
+  /*!
+   * \brief The \p DLDeviceType (represtented as an int) of the device. If \p target is known then
+   * this will be equal to \p target->kind->device_type. If \p target is null then the target is to
+   * be determined by a later pass.
+   *
+   * This is needed to support the legacy "on_device" and "device_copy" calls which only allow
+   * a \p DLDeviceTypes (as an integer) to be given.
+   *
+   * kInvalidDeviceType denotes unconstrained.
+   */
+  int device_type_int;
+
+  DLDeviceType device_type() const { return static_cast<DLDeviceType>(device_type_int); }
+
+  /*!
+   * \brief The 'virtual' device identifier for the device. This must be resolved to a physical
+   * device identifier either during compilation or at runtime.
+   *
+   * -1 denotes unconstrained.
+   */
+  int virtual_device_id;
+
+  /*!
+   * \brief The \p Target describing how to compile for the device.
+   *
+   * Null denotes unconstrained. Note that if a target later becomes known for this \p SEScope
+   * then it must be consistent with the \p device_type if that is already known. This is
+   * enforced by the Join and Default methods.
+   */
+  Target target;
+
+  /*!
+   * \brief The scope of memory within the device.
+   *
+   * Empty denotes unconstrained.
+   */
+  MemoryScope memory_scope;
+
+  /*!
+   * \brief Returns true if scope is fully unconstrained, ie no target/device type, virtual device
+   * id or memory scope is specified.
+   */
+  bool is_fully_unconstrained() const {
+    return !target.defined() && device_type() == kInvalidDeviceType && virtual_device_id == -1 &&
+           memory_scope.empty();
+  }
+
+  /*!
+   * \brief Returns true if scope is fully constrained, ie target, virtual device id and
+   * memory scope are all specified.
+   */
+  bool is_fully_constrained() const {
+    return target.defined() && virtual_device_id != -1 && !memory_scope.empty();
+  }
+
+  Device ToDevice() const {
```

Comment:
Document the behavior, note that the device id is virtual device api and may not corresponds to the real mappings(if any)

Reply:
done